clients/src/main/java/org/apache/kafka/common/protocol/types/ArrayOf.java

```diff
@@ -24,13 +24,29 @@
 public class ArrayOf extends Type {
 
     private final Type type;
+    private final boolean nullable;
 
     public ArrayOf(Type type) {
+        this(type, false);
+    }
+
+    public ArrayOf(Type type, boolean nullable) {
         this.type = type;
+        this.nullable = nullable;
+    }
+
+    @Override
+    public boolean isNullable() {
+        return nullable;
     }
 
     @Override
     public void write(ByteBuffer buffer, Object o) {
+        if (o == null) {
```

Comment:
I think all of the other types assume validate will be called before calling write and that the written object would be the one returned by the validate call.

Here is a code snippet from `Schema.write(...)`:
```scala
Object value = f.type().validate(r.get(f));
f.type.write(buffer, value);
```